File: runtime/python-agent/app/api/projects.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends

from app.deps import get_repo
from app.errors import GoyaisApiError

router = APIRouter(prefix="/v1", tags=["projects"])


@router.get("/projects")
async def list_projects(repo=Depends(get_repo)):
    return {"projects": await repo.list_projects()}


@router.post("/projects")
async def create_project(payload: dict, repo=Depends(get_repo)):
    project_id = payload.get("project_id", str(uuid.uuid4()))
    name = payload.get("name", project_id)
    workspace_path = payload["workspace_path"]
    await repo.create_project(project_id, name, workspace_path)
    event_id = await repo.insert_system_event(
        "project_upserted",
        {
            "entity": "project",
            "project_id": project_id,
            "name": name,
            "workspace_path": workspace_path,
        },
    )
    return {"project_id": project_id, "event_id": event_id}


@router.delete("/projects/{project_id}")
async def delete_project(project_id: str, repo=Depends(get_repo)):
    deleted = await repo.delete_project(project_id)
    if not deleted:
        raise GoyaisApiError(
            code="E_NOT_FOUND",
            message="project not found",
            retryable=False,
            status_code=404,
            cause="project_not_found",
        )

    event_id = await repo.insert_system_event(
        "project_deleted",
        {
            "entity": "project",
            "project_id": project_id,
        },
    )
    return {"ok": True, "event_id": event_id}

File: server/hub-server/src/routes/projects.ts
import { z } from "zod";

import type { FastifyInstance } from "fastify";

import {
  requireDomainAuth,
  requirePermission,
  requireWorkspaceIdQuery,
  requireWorkspaceMember
} from "../auth/workspace-rbac";
import type { HubDatabase } from "../db";
import { HubServerError } from "../errors";

const createProjectSchema = z.object({
  name: z.string().min(1),
  root_uri: z.string().min(1)
});

const projectParamsSchema = z.object({
  project_id: z.string().min(1)
});

interface RegisterProjectRoutesOptions {
  db: HubDatabase;
}

export function registerProjectRoutes(app: FastifyInstance, options: RegisterProjectRoutesOptions): void {
  app.get("/v1/projects", async (request) => {
    const user = requireDomainAuth(request, options.db);
    const workspaceId = requireWorkspaceIdQuery(request);
    const membership = requireWorkspaceMember(request, options.db, user, workspaceId);
    requirePermission(options.db, membership.role_id, "project:read");

    const projects = options.db.listProjects(workspaceId);
    return { projects };
  });

  app.post("/v1/projects", async (request) => {
    const user = requireDomainAuth(request, options.db);
    const workspaceId = requireWorkspaceIdQuery(request);
    const membership = requireWorkspaceMember(request, options.db, user, workspaceId);
    requirePermission(options.db, membership.role_id, "project:write");

    const parsed = createProjectSchema.safeParse(request.body);
    if (!parsed.success) {
      throw new HubServerError({
        code: "E_VALIDATION",
        message: "Invalid project payload.",
        retryable: false,
        statusCode: 400,
        details: { issues: parsed.error.issues },
        causeType: "project_payload"
      });
    }

    const project = options.db.createProject({
      workspaceId,
      name: parsed.data.name.trim(),
      rootUri: parsed.data.root_uri.trim(),
      createdBy: user.user_id
    });

    return { project };
  });

  app.delete("/v1/projects/:project_id", async (request) => {
    const user = requireDomainAuth(request, options.db);
    const workspaceId = requireWorkspaceIdQuery(request);
    const membership = requireWorkspaceMember(request, options.db, user, workspaceId);
    requirePermission(options.db, membership.role_id, "project:write");

    const paramsParsed = projectParamsSchema.safeParse(request.params);
    if (!paramsParsed.success) {
      throw new HubServerError({
        code: "E_VALIDATION",
        message: "Invalid project route params.",
        retryable: false,
        statusCode: 400,
        details: { issues: paramsParsed.error.issues },
        causeType: "project_params"
      });
    }

    const deleted = options.db.deleteProject(workspaceId, paramsParsed.data.project_id);
    if (!deleted) {
      throw new HubServerError({
        code: "E_NOT_FOUND",
        message: "Project not found.",
        retryable: false,
        statusCode: 404,
        causeType: "project_lookup"
      });
    }

    return {
      ok: true
    };
  });
}
